Sp/tuna worker interface typehint support

technicalgrp89 opened the pull request:
Worker_interface.py - type annotations/type hints support
utils.groovy - mypy command support

technicalgrp89 committed:
type hints support intermitent

technicalgrp89 committed:
type hint support - intermittent

technicalgrp89 committed:
type annotate update

technicalgrp89 committed:
yapf fix

technicalgrp89 committed:
mypy cmd correction

technicalgrp89 committed:
trial

technicalgrp89 committed:
debug codcov

technicalgrp89 committed:
code formatting

technicalgrp89 committed:
Percentile calculation change.

alexandraBara commented on tests/covscripts/coverage.py:
```diff
@@ -80,10 +80,10 @@ def branch(root_dir, coverage_file_name_txt, coverage_file_name_json):
+  if code_cov_diff > 0.5:
     raise ValueError(
-        f"Code Coverage Decreased by more than 0.5%%. Prev:{prev_cov}, cur: {cur_cov}"
+        f"Code Coverage Diffrence is Decreased by more than 0.5%%. Prev:{prev_cov}, cur: {cur_cov}"
```
```suggestion
        f"Code coverage decreased by more than 0.5. Prev:{prev_cov}, cur: {cur_cov}"
```

alexandraBara commented on tuna/worker_interface.py:
```diff
@@ -234,22 +253,31 @@ def refresh_query_objects(self, session, rows):
+    job_set_attr: List[str]
+    session: DbSession
+    ids: list
+    row: str
```
its not a string, its of JobMixin type i believe. It is a row in the Job table which is an implementation of the JobMixin interface

alexandraBara commented on tuna/worker_interface.py:
```diff
@@ -265,14 +293,14 @@ def get_job(self, find_state, set_state, imply_end):
-              for job in job_tables:
-                job.state = set_state
+              for self.job in job_tables:
+                self.job.state = set_state  #type: ignore
```
state is an enum, why is it ignored?

alexandraBara commented on tuna/worker_interface.py:
```diff
@@ -454,10 +490,10 @@ def reset_job_state(self):
         break
 
-  def run(self):
+  def run(self) -> bool:  #type: ignore
```
why is there a type ignore here?

technicalgrp89 replied:
run() return type is "None" in supertype "BaseProcess"

technicalgrp89 committed:
conflicts resolved - latest code

technicalgrp89 committed:
review comments update

technicalgrp89 committed:
Job tables definitions added, review comments update

cderb commented on tests/covscripts/coverage.py:
```diff
@@ -80,10 +80,10 @@ def branch(root_dir, coverage_file_name_txt, coverage_file_name_json):
+  if code_cov_diff > 0.5:
     raise ValueError(
-        f"Code Coverage Decreased by more than 0.5%%. Prev:{prev_cov}, cur: {cur_cov}"
+        f"Code Coverage Decreased by more than 0.5. Prev:{prev_cov}, cur: {cur_cov}"
```
Lost the % sign?

cderb commented on tuna/worker_interface.py:
```diff
@@ -87,31 +94,32 @@ def __init__(self, **kwargs):
     connect_db()
 
-    self.job = SimpleDict()
+    self.job: JobMixin = SimpleDict()
```
Same fields in practice as JobMixin, but this is a SimpleDict, as can be seen by the constructor. SimpleDict has no ancestry with JobMixin.

cderb replied:
this

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,32 +219,34 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
     return True
 
-  def get_job_from_tuple(self, job_tuple):
+  def get_job_from_tuple(self, job_tuple: str) -> Union[str, JobMixin, None]:
```
return Optional[SimpleDict]

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,32 +219,34 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
-  def get_job_from_tuple(self, job_tuple):
+  def get_job_from_tuple(self, job_tuple: str) -> Union[str, JobMixin, None]:
     """find job table in a job tuple"""
+    tble: str
```
not a string, SimpleDict

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,32 +219,34 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
-  def get_job_tables(self, job_rows):
+  def get_job_tables(
+      self, job_rows: List[Tuple[SimpleDict,
+                                 ...]]) -> List[Tuple[SimpleDict, ...]]:
```
return List[SimpleDict]

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,32 +219,34 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
     """find job tables in query results"""
     if has_attr_set(job_rows[0], self.job_attr):
-      job_tables = job_rows
+      job_tables: List[Tuple[SimpleDict, ...]] = job_rows
```
List[SimpleDict]

cderb replied:
this

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,32 +219,34 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
     else:
-      job_i = 0
+      job_i: int = 0
+      tble: str
```
SimpleDict

cderb replied:
this
technically we don't know because "..." is in the tuple, but SimpleDict is good enough.

cderb replied:
tble is SimpleDict, not a string

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
         session.refresh(obj_tuple)
 
-  def job_queue_push(self, job_rows):
+  def job_queue_push(self, job_rows: List[Tuple[JobMixin, SimpleDict]]) -> None:
```
List[Tuple[SimpleDict, ...]]

technicalgrp89 replied:
List[SimpleDict] changed to this.

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
-  def job_queue_push(self, job_rows):
+  def job_queue_push(self, job_rows: List[Tuple[JobMixin, SimpleDict]]) -> None:
     """load job_queue with info for job ids"""
+    job: JobMixin
```
SimpleDict

cderb replied:
this

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
+  def job_queue_push(self, job_rows: List[Tuple[JobMixin, SimpleDict]]) -> None:
     """load job_queue with info for job ids"""
+    job: JobMixin
+    job_tuple: JobMixin
```
Tuple[SimpleDict, ...]

cderb replied:
this

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
       self.job_queue.put(job_tuple)
       job = self.get_job_from_tuple(job_tuple)
-      self.logger.info("Put job %s %s %s", job.id, job.state, job.reason)
+      if job is not None:
```
Condition not necessary, unreachable.

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
-  def get_job(self, find_state, set_state, imply_end):
+  def get_job(self, find_state: str, set_state: str, imply_end: bool) -> bool:
     """Interface function to get new job for builder/evaluator"""
+    job_rows: JobMixin
```
List[Tuple[SimpleDict, ...]]

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
+  def get_job(self, find_state: str, set_state: str, imply_end: bool) -> bool:
     """Interface function to get new job for builder/evaluator"""
+    job_rows: JobMixin
+    job_tables: JobMixin
```
List[SimpleDict]

cderb commented on tuna/worker_interface.py:
```diff
@@ -234,22 +255,33 @@ def refresh_query_objects(self, session, rows):
+    job_set_attr: List[str]
+    session: DbSession
+    ids: list
+    row: JobMixin
```
SimpleDict

cderb commented on tuna/worker_interface.py:
```diff
@@ -267,12 +299,12 @@ def get_job(self, find_state, set_state, imply_end):
               self.logger.info("%s jobs %s", find_state, ids)
-              for job in job_tables:
-                job.state = set_state
+              for self.job in job_tables:
+                self.job.state = set_state
```
Don't add self. this severely changes functionality.

cderb commented on tuna/worker_interface.py:
```diff
@@ -267,12 +299,12 @@ def get_job(self, find_state, set_state, imply_end):
                 job_set_attr = ['state']
-                query = gen_update_query(job, job_set_attr,
-                                         self.dbt.job_table.__tablename__)
+                query: str = gen_update_query(self.job, job_set_attr,
```
don't add self

cderb reviewed (changes requested):
Needs work. Necessary changes not limited to comments.

technicalgrp89 committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/MITunaX into sp/worker_interface

technicalgrp89 committed:
definitions related to Job table-update

technicalgrp89 committed:
CI error fix

technicalgrp89 committed:
code review update

cderb commented on tuna/worker_interface.py:
```diff
@@ -338,66 +368,73 @@ def callback():
     return ret_code, out, err
 
-  def exec_docker_cmd(self, cmd):
+  def exec_docker_cmd(self, cmd: str) -> Tuple[int, str, ChannelStderrFile]:
```
-> Tuple[int, StringIO, StringIO]

technicalgrp89 replied:
Tuple[int, str, StringIO] changed

cderb commented on tuna/worker_interface.py:
```diff
@@ -338,66 +368,73 @@ def callback():
+  def exec_command(self, cmd: str) -> Tuple[int, StringIO, StringIO]:
     """execute on native machine"""
     ret_code, out, err = self.cnx.exec_command(cmd, timeout=LOG_TIMEOUT)
     if err is not None and hasattr(err, 'channel'):
       self.logger.info(err)
       err.channel.settimeout(LOG_TIMEOUT)
```
delete this if statement

cderb commented on tuna/worker_interface.py:
```diff
@@ -338,66 +368,73 @@ def callback():
       self.logger.info('Error executing docker cmd: %s \n err: %s', cmd,
                        err.read())
 
     if err is not None and hasattr(err, 'channel'):
       err.channel.settimeout(LOG_TIMEOUT)
       self.logger.info(err)
```
delete this if statement

cderb commented on tuna/worker_interface.py:
```diff
@@ -503,17 +541,20 @@ def run(self):
     """Run cmd and return ret_code"""
+    ret_code: int
+    out: str
+    err: ChannelStderrFile
```
StringIO

technicalgrp89 committed:
last changes reverted to avoid StringIO.translate err

technicalgrp89 committed:
ChannelStdErr types removed

technicalgrp89 committed:
CI fix

cderb commented on tuna/worker_interface.py:
```diff
@@ -338,66 +367,71 @@ def callback():
       assert session_retry(session, callback, lambda x: x(), self.logger)
 
-  def exec_command(self, cmd):
+  def exec_command(self, cmd: str) -> Tuple[int, StringIO, StringIO]:
```
Would be good if the returns from this value were of consistent type with exec_docker_cmd.

technicalgrp89 replied:
When we normalize the return parameters of exec_docker_cmd with exec_command, we get an error on _StringIO object while processing string. As _StringIO doesn't support translate() cmd( in sqlalchemy lib). get_rocm_v->exe_docker_cmd-> returns error, when we set the version parameter as StringIO. As of now, We need to define the return parameters of exec_docker_cmd as Tuple[int, str, StringIO].

cderb replied:
It would do just as well to match this functions output types to exec_docker_cmd, eg use Tuple[int, str, StringIO] and match the string parsing functionality of the other function.

cderb commented on tuna/worker_interface.py:
```diff
@@ -454,10 +488,10 @@ def reset_job_state(self):
         break
 
-  def run(self):
+  def run(self) -> Optional[bool]:  #type: ignore[override]
```
why ignore?

technicalgrp89 replied:
run() return type is "None" in supertype "BaseProcess"

technicalgrp89 committed:
JobMixin->SimpleDict changes

technicalgrp89 committed:
conds->cmds variable rename

cderb commented on tuna/miopen/worker/fin_class.py:
```diff
@@ -146,17 +146,16 @@ def set_db_tables(self):
     ret = []
     if self.fin_steps:
-      conds.append(f"fin_step like '%{self.fin_steps[0]}%'")
+      cmds.append(f"fin_step like '%{self.fin_steps[0]}%'")
```
Please keep as conds = conditions, This is the where clause for an sql query.

cderb commented on tuna/worker_interface.py:
```diff
@@ -142,53 +159,54 @@ def set_logger(self, logger_name):
-    if cond_str:
-      cond_str = f"WHERE {cond_str}"
-    cond_str += f" ORDER BY retries ASC LIMIT {self.claim_num} FOR UPDATE"
+    cmd_str = ' AND '.join(cmds)
```
Please keep as cond_str, this is a query clause.

cderb commented on tuna/worker_interface.py:
```diff
@@ -142,53 +159,54 @@ def set_logger(self, logger_name):
 
   def get_job_objs(self, session: DbSession,
-                   find_state: str) -> List[Tuple[SimpleDict, ...]]:
+                   find_state: str) -> List[SimpleDict]:
```
Return type is incorrect, see return of compose_work_objects. This will depend on implementation, but at this level generically: List[Tuple[SimpleDict, ...]]

cderb commented on tuna/worker_interface.py:
```diff
@@ -142,53 +159,54 @@ def set_logger(self, logger_name):
+                   find_state: str) -> List[SimpleDict]:
     """Get list of job objects"""
-    conds = [f"session={self.dbt.session.id}", "valid=1"]
+    entries: List[SimpleDict]
```
Incorrect typing.

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,56 +218,56 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
     return True
 
-  def get_job_from_tuple(self, job_tuple):
+  def get_job_from_tuple(self, job_tuple: SimpleDict) -> Optional[SimpleDict]:
```
incorrect typing
purpose of this function is to extract the SimpleDict - job object from job_tuple
job_tuple in this case is assumed to either be Tuple[SimpleDict, ...] or SimpleDict

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,56 +218,56 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
-        session.refresh(obj_tuple)
-
-  def job_queue_push(self, job_rows):
+  def job_queue_push(self, job_rows: List[SimpleDict]) -> None:
```
incorrect type 
job_rows is List[Tuple[SimpleDict, ...]]

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,56 +218,56 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
+  def job_queue_push(self, job_rows: List[SimpleDict]) -> None:
     """load job_queue with info for job ids"""
+    job: SimpleDict
+    job_tuple: SimpleDict
```
job_tuple is Tuple[SimpleDict, ...]

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,56 +218,56 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
-  def get_job(self, find_state, set_state, imply_end):
+  def get_job(self, find_state: str, set_state: str, imply_end: bool) -> bool:
     """Interface function to get new job for builder/evaluator"""
+    job_rows: List[SimpleDict]
```
List[Tuple[SimpleDict, ...]]

technicalgrp89 committed:
type definitions update
fin_class.py changes reverted
exec_cmd,exec_docker_cmd return normalized
get_rocm_v(),get_miopen_v() type definitions are changed
ret_code,out,err parameter values checks added

cderb commented on tuna/worker_interface.py:
```diff
@@ -171,24 +188,25 @@ def compose_work_objs(self, session: DbSession,
     """Get list of job objects"""
-    conds = [f"session={self.dbt.session.id}", "valid=1"]
+    entries: List[Tuple[SimpleDict, ...]]
+    cmds: list = [f"session={self.dbt.session.id}", "valid=1"]
```
don't change this var name: conds

cderb commented on tuna/worker_interface.py:
```diff
@@ -200,56 +218,57 @@ def check_jobs_found(self, job_rows, find_state, imply_end):
     return None
 
-  def get_job_tables(self, job_rows):
+  def get_job_tables(self, job_rows: List[SimpleDict]) -> List[SimpleDict]:
```
job_rows: List[Tuple[SimpleDict, ...]]

cderb commented on tuna/worker_interface.py:
```diff
@@ -322,82 +342,92 @@ def set_job_state(self,
+      self.logger.info('Error executing exec cmd: %s \n err: %s', cmd,
+                       err.read())
+
+    return ret_code, line, err
```
if not out, then line isn't defined.

cderb commented on tuna/worker_interface.py:
```diff
@@ -322,82 +342,92 @@ def set_job_state(self,
-      err.channel.settimeout(LOG_TIMEOUT)
-      self.logger.info(err)
-    return ret_code, out, err
+    return ret_code, line, err
```
if not out, then line isn't defined.

cderb commented on tuna/worker_interface.py:
```diff
@@ -454,10 +484,10 @@ def reset_job_state(self):
         break
 
-  def run(self):
+  def run(self) -> bool:  #type: ignore[override]
```
Add a note that that this is due to the parent class typing.

cderb reviewed (changes requested):
Some previous comments not yet addressed.

cderb committed:
Update worker_interface.py

tidy function

cderb commented on tuna/worker_interface.py:
```diff
@@ -503,17 +545,20 @@ def run(self):
+          err_str: str = err.read()
           self.logger.error('%s : %s', ret_code, err_str)
-          if "disk I/O error" in err_str:
+          if "disk I/O error" in cast(str, err_str):
```
why does this need a cast?

cderb committed:
Update worker_interface.py

update comment

technicalgrp89 committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/MITunaX into sp/worker_interface

technicalgrp89 committed:
format code

technicalgrp89 committed:
local changes

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -67,17 +72,17 @@ def __init__(self, **kwargs):
     super().__init__()
 
-    allowed_keys = set([
+    allowed_keys: Set = set([
```
Why not just `{'machine', 'gpu_id', 'num_procs', 'barred', 'bar_lock', 'envmt', 'reset_interval', 'job_queue', 'job_queue_lock', 'result_queue', 'result_queue_lock', 'label', 'fetch_state', 'end_jobs', 'session_id'}`. Also, perhaps change the typehint to `Set[str]`

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -171,7 +187,8 @@ def compose_work_objs(self, session: DbSession,
     """Get list of job objects"""
-    conds = [f"session={self.dbt.session.id}", "valid=1"]
+    entries: List[Tuple[SimpleDict, ...]]
+    conds: list = [f"session={self.dbt.session.id}", "valid=1"]
```
`conds: List[str]`?

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -525,6 +563,5 @@ def run_command(self, cmd):
 
     if ret_code != 0:
-      ret_code = None
-
+      ret_code = 0
```
I could be missing something, but if the command errored out, why would you want to manually override the return code to 0?

technicalgrp89 replied:
ret_code != 0 
ret_code =  0 lines are removed , as ret_code is retuned as received

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -30,25 +30,30 @@
+from tuna.connection import Connection
 from tuna.utils.utility import SimpleDict
 
 MAX_JOB_RETRIES = 10
```
Instead of leaving MAX_JOB_RETRIES, LOG_TIMEOUT in the global scope, I'd say move them to inside the WorkerInterface class

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -67,17 +72,17 @@ def __init__(self, **kwargs):
         'reset_interval', 'job_queue', 'job_queue_lock', 'result_queue',
         'result_queue_lock', 'label', 'fetch_state', 'end_jobs', 'session_id'
     ])
     self.__dict__.update((key, None) for key in allowed_keys)
```
Why need line 80 when you have lines 83-98? Also, modifying `__dict__`  might perhaps best be avoided

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -87,31 +92,32 @@ def __init__(self, **kwargs):
+    self.session_id: int = None
 
     self.__dict__.update(
         (key, value) for key, value in kwargs.items() if key in allowed_keys)
```
why not use `setattr` ?

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -67,17 +72,17 @@ def __init__(self, **kwargs):
     #multiprocess vars
-    self.gpu_id = None
+    self.gpu_id: int = None
     self.num_procs = None
```
Missing type hints

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -87,31 +92,32 @@ def __init__(self, **kwargs):
+    self.job: SimpleDict = SimpleDict()
+
     try:
       self.job_attr = [column.name for column in inspect(self.dbt.job_table).c]
```
`self.job_attr: List[str]`?

msaudulhassan commented on tuna/worker_interface.py:
```diff
@@ -303,12 +321,14 @@ def get_job(self, find_state, set_state, imply_end):
     return False
 
-  # JD: This should take a session obj as an input to remove the creation of an extraneous session
+  # This should take a session obj as an input to remove the creation of an extraneous session
```
Mark it as TODO so it's easier to search for todo's later on

technicalgrp89 committed:
__dict__.update replaced by setattr

technicalgrp89 committed:
latest code merge

technicalgrp89 committed:
reverting __dict__ changes

msaudulhassan reviewed (approved):
You might want to have @cderb take another look at this, especially at the changes you made in lieu of:
https://github.com/ROCmSoftwarePlatform/MITuna/pull/903#discussion_r1253633327
https://github.com/ROCmSoftwarePlatform/MITuna/pull/903#discussion_r1253629717
https://github.com/ROCmSoftwarePlatform/MITuna/pull/903#discussion_r1248291399
